docs/api_design.md

```diff
@@ -0,0 +1,69 @@
+# API Design
+
+> Warning: This is work in progress and APIs are very likely to change.
+
+This is a working draft of the API design for **s2-api**.
+
+## Anatomy
+
+```
+s2_api
+|-- Bot
+    |-- Action
+    |-- `tick()`
+|-- Connection
+|-- Engine
+    |-- GameState
+    |-- `run()`
+|-- Remote
+    |-- Controller
+        |-- `act()`, `observe()`, `ping()`, ...
+    |-- Protocol
+        |-- `send()` and `receive()`
+|-- SC2API
+    |-- Request/Response
+        |-- (protocol types)
+```
+
+## Bot
+
+The agent used to infer **GameState** and take an **Action**. The **Bot** will
+be notified of **GameState** via a `tick()`.
+
+### Action
+
+In-game actions that can be performed by a **Bot**.
+
+## Connection
+
+Responsible for establishing the WebSocket connection to a StarCraft II running
+process for sending and receiving messages. Used by the **Remote::Protocol**.
```

Comment:
We may have an issue here in that there needs to be two-way communication between `Connection` and the websocket client. Therefore you're either going to need two channels, or will have to do something fancy with callbacks.

I'll have a dig through the C++ API and see how they deal with it...